nexus/sdk-python.md

```diff
@@ -0,0 +1,330 @@
+    hello: nexusrpc.Operation[HelloInput, HelloOutput]
+```
+
+Note that:
+
+- What the user actually writes here defines the operations as purely type-level constructs: the decorator creates `NexusOperation` instances at import time that can be referenced by the caller-side Nexus service client.
+
+The interface will be used by the caller as follows:
+
+```python
+from my_nexus_service import EchoInput, EchoOutput, MyNexusService
+
+@workflow.defn
+class EchoCallerWorkflow:
+    def __init__(self):
+        self.nexus_client = NexusClient(
+            MyNexusService,  # or string name "my-nexus-service",
+            "my-nexus-endpoint-name",
+            schedule_to_close_timeout=timedelta(seconds=30),
+        )
+
+    @workflow.run
+    async def run(self, message: str) -> EchoOutput:
+        return await self.nexus_client.execute_operation(
+            MyNexusService.echo,
+            EchoInput(message),
+        )
+
+```
+
+**[ALTERNATIVES CONSIDERED]** One might consider using a traditional (`typing.Protocol` / `abc.ABC`) interface (as Java does). However, while the caller-side Nexus service client might conform to those signatures, the handler-side service implementation would not (see below, but for at least two reasons: (1) operations are not represented by their start methods but rather by nullary factory functions returning an operation instance, and (2) even if the handler-side signature were that of the start method, it contains an option container from the request that would be extraneous in an interface definition).
+
+# 2. Implementing a Nexus operation
+
+Traditionally, in order to define an RPC/HTTP handler, you specify a function to act as the handler, and you arrange for a service or operation instance with a lifetime equal to that of the server, containing state created at server-start time, to be in-scope in the function body. For Python examples, see [gRPC,](https://github.com/grpc/grpc/blob/4e9357bca1408596663a218c0c608a4c0560a867/examples/python/route_guide/route_guide_server.py#L65) [xmlrpc [stdlib]](https://docs.python.org/3/library/xmlrpc.server.html#simplexmlrpcserver-example), [starlette/FastAPI](https://www.starlette.io/applications/#storing-state-on-the-app-instance), [Django class-based-views](https://docs.djangoproject.com/en/5.1/topics/class-based-views/). So, a Python developer writing an RPC/HTTP handler will expect to write something like
+
+```python
+# This code (from Python gRPC docs) is not part of the proposal
+class MyService:  # instantiated at server-start time
+    def my_handler(self, my_input, context) -> MyResult: ...
+```
+
+However, unlike traditional RPC/HTTP handlers, a Nexus operation represents a long-running execution and must support three additional verbs (`get_info`, `get_result`, `cancel`), and must support `start` returning both synchronously and asynchronously (including making that decision at request-handling time).
+
+## 2a. Fully manual class-based operation definition
```

Comment:
Can you show an example of defining a Nexus service dynamically at runtime? We took care in Java to only make the annotation/decorator-based forms be sugar over the manual definitions that could be created (for the services, operations, service impls, and operation handlers). You don't _have_ to use the annotation/decorator-based forms if you need to build at runtime dynamically.